src/include/daos_fs_sys.h

```diff
@@ -0,0 +1,581 @@
+}
+
+/**
+ * Compare hash entry key.
+ * Simple string comparison of name.
+ */
+static bool
+dfs_sys_hash_key_cmp(struct d_hash_table *table, d_list_t *rlink,
+		     const void *key, unsigned int ksize)
+{
+	struct dfs_sys_hash_hdl	*hdl = dfs_sys_hash_hdl_obj(rlink);
+
+	return (strcmp(hdl->name, (const char *)key) == 0);
+}
+
+/**
+ * Add reference to hash entry.
+ */
+static void
+dfs_sys_hash_rec_addref(struct d_hash_table *htable, d_list_t *rlink)
+{
+	struct dfs_sys_hash_hdl *hdl = dfs_sys_hash_hdl_obj(rlink);
+
+	hdl->num_refs++;
+}
+
+/**
+ * Decrease reference to hash entry.
+ */
+static bool
+dfs_sys_hash_rec_decref(struct d_hash_table *htable, d_list_t *rlink)
+{
+	struct dfs_sys_hash_hdl *hdl = dfs_sys_hash_hdl_obj(rlink);
+	bool			no_refs;
+
+	D_ASSERT(hdl->num_refs > 0);
+
+	hdl->num_refs--;
+	no_refs = (hdl->num_refs == 0);
+
+	return no_refs;
+}
+
+/*
+ * Free a hash entry.
+ */
+static void
+dfs_sys_hash_rec_free(struct d_hash_table *htable, d_list_t *rlink)
+{
+	struct dfs_sys_hash_hdl *hdl = dfs_sys_hash_hdl_obj(rlink);
+
+	/* TODO is the assert necessary? */
+	D_ASSERT(d_hash_rec_unlinked(&hdl->entry));
+	dfs_sys_hash_hdl_delete(hdl);
+}
+
+/**
+ * Operations for the hash table.
+ */
+static d_hash_table_ops_t dfs_sys_hash_hdl_ops = {
+	.hop_key_cmp	= dfs_sys_hash_key_cmp,
+	.hop_rec_addref = dfs_sys_hash_rec_addref,
+	.hop_rec_decref	= dfs_sys_hash_rec_decref,
+	.hop_rec_free	= dfs_sys_hash_rec_free
+};
+
+
+/**
+ * Mount a file system with dfs_mount and optionally initialize the hash.
+ */
+/* TODO to be more consistent with DFS, maybe this header */
+// dfs_sys_mount(daos_handle_t poh, daos_handle_t coh, int flags, dfs_sys_t **_dfs_sys)
+/* TODO maybe let a different flags be passed for dfs_sys hashing/locking/future */
+int
+dfs_sys_mount(dfs_sys_t *dfs_sys, daos_handle_t *poh, daos_handle_t *coh,
+	      int flags, bool use_hash)
+{
+	int rc;
+
+	rc = dfs_mount(*poh, *coh, flags, &dfs_sys->dfs);
+	if (rc) {
+	D_ERROR("dfs_mount() failed (%d)\n", rc);
+		D_GOTO(out, rc);
+	}
+
+	/** TODO make this thread safe */
+	/** TODO allow size to be passed in? Default size? */
+	if (use_hash) {
+		rc = d_hash_table_create(D_HASH_FT_NOLOCK, 16, NULL,
+					 &dfs_sys_hash_hdl_ops,
+					&dfs_sys->dfs_hash);
+		if (rc) {
+			/** TODO error handling */
+			D_GOTO(err_hash, rc);
+		}
+	}
+
+err_hash:
+	dfs_umount(dfs_sys->dfs);
+out:
+	return rc;
+}
+
+/**
+ * Unmount a file system with dfs_mount and destroy the hash.
+ */
+int
+dfs_sys_umount(dfs_sys_t *dfs_sys)
+{
+	int rc;
+
+	/** TODO check params to this function */
+	rc = d_hash_table_destroy(dfs_sys->dfs_hash, true);
+	/** TODO error handling */
+
+	rc = dfs_umount(dfs_sys->dfs);
+	/** TODO error handling */
+
+	return rc;
+}
+
+
+/**
+ * Try to get name from the hash.
+ * If not found, call dfs_lookup on name
+ * and store in the hash.
+ * Stores the hashed obj in _obj.
+ */
+static int
+dfs_sys_hash_lookup(const char *name, dfs_sys_t *dfs_sys, dfs_obj_t **_obj)
```

Comment:
Any function that takes a char * should take a len as well, and this applies to public as well as internal interfaces.